src/core/column/qcut.h

```diff
@@ -0,0 +1,178 @@
+//------------------------------------------------------------------------------
+// Copyright 2020 H2O.ai
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#ifndef dt_COLUMN_CUT_h
+#define dt_COLUMN_CUT_h
+#include <memory>
+#include "column/const.h"
+#include "column/virtual.h"
+#include "ltype.h"
+#include "models/utils.h"
+#include "parallel/api.h"
+#include "sort.h"
+
+namespace dt {
+
+/**
+  *  Virtual column to bin numeric values into equal-population
+  *  discrete intervals, i.e. quantiles. In reality, for some data
+  *  these quantiles won't have exactly the same population.
+  *
+  *  Quantiles are generated based on the element/group information
+  *  obtained from the groupby operation, i.e. rowindex and offsets.
+  *  These groups, having ids 0, 1, ..., ngroups - 1, are binned
+  *  into `nquantiles` equal-width discrete intervals. As a result,
+  *  all the duplicates of a value `x` will go to the same `x_q` quantile.
+  *
+  *  In a simple case, when all the values are missing or constant,
+  *  `Qcut_ColumnImpl::make()` will immediately return a pointer to the
+  *  corresponding virtual column implementation: `ConstNa_ColumnImpl`
+  *  or `ConstInt_ColumnImpl`.
+  *
+  *  In the general case `Qcut_ColumnImpl::make()` will return a pointer
+  *  to the `Latent_ColumnImpl` that wraps a `Qcut_ColumnImpl*`.
+  *  This column implementation will be materialized on the first access
+  *  to the data, because it is much more efficient to generate
+  *  all the quantiles at once in parallel.
+  *
+  */
+class Qcut_ColumnImpl : public Virtual_ColumnImpl {
+  private:
+    Column col_;
+    RowIndex ri_;
+    Groupby gb_;
+    int32_t nquantiles_;
+    size_t : 32;
+
+  public:
+    static ColumnImpl* make(Column&& col, int32_t nquantiles) {
+      xassert(nquantiles > 0);
+      auto res = group({col}, {SortFlag::NONE});
```

Comment:
sorting/grouping is a quite expensive operation, so it would be better to hide inside the `Qcut_ColumnImpl`, so that it would get materialized only when it's actually needed.

Reply:
Yes, but then we have this problem returning a virtual column from the `Latent_ColumnImpl`.

Reply:
You could always just do `.materialize()` on that column before returning.
Plus, we discussed yesterday how to enable virtual columns with Latent column: create a special "placement clone" virtual method.

Reply:
Yes, I first had materialization turned on for these columns, but didn't like that solution. May be let me revert it back, as it is still more "optimal" than doing a groupby.

As for the Latent column improvements, I propose we do it in a separate PR along with some other changes to `materialize` API.